### PR TITLE
drivers: watchdog: Add SiLabs Gecko WDOG driver

### PR DESCRIPTION
**Pull request by Oanerer**
This PR is an initial attempt at adding a watchdog driver for the Silicon Labs Gecko series.

Some of the functionality of the watchdog API is not entirely clear to me, particularly the window functionality, so I would very much like to hear from someone in the know on how I should proceed or that my current implementation is how it was intended to be used.

The driver is tested on the Pearl Gecko starter kit using the default Zephyr watchdog test. The driver should be functional on most Gecko boards, however I did not add them yet as I cannot test it for myself.

### CODE REVIEW

**Comment by galak** [comment]
There are a few changes in #20321 that should get reflected in this PR.

File: soc/arm/silabs_exx32/efm32pg12b/dts_fixup.h [hunk 1]
```diff
@@ -26,4 +26,7 @@
 #define DT_GPIO_GECKO_PORTE_NAME	DT_SILABS_EFM32_GPIO_PORT_4000A0C0_LABEL
 #define DT_GPIO_GECKO_PORTF_NAME	DT_SILABS_EFM32_GPIO_PORT_4000A0F0_LABEL
 
+#define DT_WDT_0_NAME				DT_INST_0_SILABS_GECKO_WDOG_LABEL
```
Comment (galak) [inline soc/arm/silabs_exx32/efm32pg12b/dts_fixup.h]: Remove these defines, we should utilize aliases instead.

File: dts/arm/silabs/efm32pg12b.dtsi [hunk 2]
```diff
@@ -182,6 +182,22 @@
 				#gpio-cells = <2>;
 			};
 		};
+
+		wdog0: wdog@40052000 {
```
Comment (galak) [inline dts/arm/silabs/efm32pg12b.dtsi]: Add an alias for 'watchdog-0' & `watchdog-1` in this file.

**Review by galak — CHANGES_REQUESTED** [review]
Some minor dt changes to have things in sync with PR #20321

File: soc/arm/silabs_exx32/efm32pg12b/dts_fixup.h (hunk 1 above)
Comment (Oanerer) [reply soc/arm/silabs_exx32/efm32pg12b/dts_fixup.h]: Removed

File: dts/arm/silabs/efm32pg12b.dtsi (hunk 2 above)
Comment (Oanerer) [reply dts/arm/silabs/efm32pg12b.dtsi]: Added 2 aliases

**Comment by Oanerer** [comment]
> 
> 
> Some minor dt changes to have things in sync with PR #20321

Hi Galak, thanks for your review! I corrected the items in your change request.

With regards to the CODEOWNERS file: no other codeowner information regarding watchdog drivers seems to be there or am I missing something?

File: dts/arm/silabs/efm32pg12b.dtsi [hunk 3]
```diff
@@ -10,6 +10,11 @@
 #include "gpio_gecko.h"
 
 / {
+	aliases {
```
Comment (chrta) [inline dts/arm/silabs/efm32pg12b.dtsi]: These aliases are not sufficient to make the watchdog test run. I had to change the source code of the test:

~~~diff
diff --git a/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c b/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c
index d2ffb38f3f..9bfe29a8d7 100644
--- a/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c
+++ b/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c
@@ -68,7 +68,7 @@
 #ifdef CONFIG_WWDG_STM32
 #define WDT_DEV_NAME DT_WWDT_0_NAME
 #else
-#define WDT_DEV_NAME DT_WDT_0_NAME
+#define WDT_DEV_NAME DT_ALIAS_WATCHDOG_0_LABEL
 #endif
 #endif
~~~
@galak What is the proper fix for that?

File: boards/arm/efm32pg_stk3402a/efm32pg_stk3402a.yaml [hunk 4]
```diff
@@ -12,6 +12,7 @@ supported:
   - i2c
   - gpio
   - nvs
+  - watchdog
```
Comment (chrta) [inline boards/arm/efm32pg_stk3402a/efm32pg_stk3402a.yaml]: Please also add the watchdog to the board documentation index.rst file, see https://github.com/chrta/zephyr/commit/f75080cda208e0cfd70966d43ef0c4eabdd0826b

**Review by chrta — CHANGES_REQUESTED** [review]
I verified the functionality on the efr32mg board. If you want, you can cherry-pick https://github.com/chrta/zephyr/commit/f75080cda208e0cfd70966d43ef0c4eabdd0826b

After the code change mentioned in my comment, the test is working fine.

File: dts/arm/silabs/efm32pg12b.dtsi (hunk 3 above)
Comment (Oanerer) [reply dts/arm/silabs/efm32pg12b.dtsi]: I think @galak first needs to have PR [#20321](https://github.com/zephyrproject-rtos/zephyr/pull/20321) merged in.

File: boards/arm/efm32pg_stk3402a/efm32pg_stk3402a.yaml (hunk 4 above)
Comment (Oanerer) [reply boards/arm/efm32pg_stk3402a/efm32pg_stk3402a.yaml]: Added the watchdog entry to the documentation of the Pearl Gecko starter kit

**Review by Oanerer — COMMENTED** [review]
> 
> 
> I verified the functionality on the efr32mg board. If you want, you can cherry-pick [chrta@f75080c](https://github.com/chrta/zephyr/commit/f75080cda208e0cfd70966d43ef0c4eabdd0826b)
> 
> After the code change mentioned in my comment, the test is working fine.

Awesome Christian, thanks for testing and your efr32mg patch! I've added it to this PR.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

File: dts/arm/silabs/efm32pg12b.dtsi (hunk 3 above)
Comment (chrta) [reply dts/arm/silabs/efm32pg12b.dtsi]: Yes, i missed that.

**Review by chrta — APPROVED** [review]

**Comment by Oanerer** [comment]
Added CODEOWNERS information.
Corrected aliases for EFM32PG and EFR32MG from `watchdog-N` to `watchdogN` to be in line with PR [#20321](https://github.com/zephyrproject-rtos/zephyr/pull/20321)

File: drivers/watchdog/wdt_gecko.c [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
+
+	data->wdog_config = init_defaults;
+
+	data->wdog_config.perSel = (WDOG_PeriodSel_TypeDef)
+		wdt_gecko_convert_timeout(cfg->window.max);
+
+	if (cfg->window.min) {
```
Comment (mnkp) [inline drivers/watchdog/wdt_gecko.c]:
```
	if (cfg->window.min > 0) {
```
I believe it's not yet a must for the drivers but we're leaning towards writing Misra-C compliant code. At least follow few of the rules that make the code more readable.

File: drivers/watchdog/Kconfig.gecko [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+# Copyright (c) 2019 Interay Solutions B.V.
+# Copyright (c) 2019 Oane Kingma
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config WDT_GECKO
+	bool "Gecko series Watchdog (WDOG) Driver"
+	depends on SOC_FAMILY_EXX32
+	select SOC_GECKO_WDOG
+	select HAS_DTS_WDT
```
Comment (galak) [inline drivers/watchdog/Kconfig.gecko]: you can remove `select HAS_DTS_WDT`

**Review by galak — COMMENTED** [review]
Looks good, one minor nit in Kconfig.

**Review by mnkp — COMMENTED** [review]
@Oanerer could I ask you to add support for the remaining 3 SiLabs SoCs? Watchdog is one of those peripherals that will be in a high demand. I personally would like to use it on efr32fg1p SoC / efr32_slwstk6061a board. If that's OK.

File: drivers/watchdog/wdt_gecko.c [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	} else {
+		/* Normal mode */
+		data->wdog_config.winSel = wdogIllegalWindowDisable;
+	}
+
+	/* Using fixed 75% value for early warning interrupt */
+	data->wdog_config.warnSel = wdogWarnTime75pct;
+
+	/* Set mode of watchdog and callback */
+	switch (cfg->flags) {
```
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/watchdog/wdt_gecko.c]: Sorry, the way this is currently written is correct, no need to change it.

File: drivers/watchdog/Kconfig.gecko (hunk 6 above)
Comment (Oanerer) [reply drivers/watchdog/Kconfig.gecko]: Removed, thanks!

**Comment by Oanerer** [comment]
> 
> 
> @Oanerer could I ask you to add support for the remaining 3 SiLabs SoCs? Watchdog is one of those peripherals that will be in a high demand. I personally would like to use it on efr32fg1p SoC / efr32_slwstk6061a board. If that's OK.

Hi Piotr, I've added support for the efr32_slwstk6061a board as per your request. The other 2 boards do not yet support interrupts for the watchdog, so that also requires a modification to the driver. As I do not have access to one of those boards and therefore cannot test it, I will not add support for them now.

File: drivers/watchdog/wdt_gecko.c [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+			wdt_gecko_convert_window(cfg->window.min,
+				timeout_in_cycles_tbl[
+				data->wdog_config.perSel]);
+	} else {
+		/* Normal mode */
+		data->wdog_config.winSel = wdogIllegalWindowDisable;
+	}
+
+	/* Using fixed 75% value for early warning interrupt */
+	data->wdog_config.warnSel = wdogWarnTime75pct;
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Ah thanks, that clears up a few things for me! I modified the driver to only support the TOUT interrupt, removed the warning setting and interrupt altogether and modified the windows setting to be configurable up to 87.5% of the timeout period. If you see any other issues, please let me know.

File: drivers/watchdog/wdt_gecko.c [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+	u32_t flags;
+
+	/* Clear IRQ flags */
+	flags = WDOGn_IntGet(wdog);
+	WDOGn_IntClear(wdog, flags);
+
+	if (data->callback) {
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Changed

File: drivers/watchdog/wdt_gecko.c (hunk 5 above)
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Thanks for pointing it out, I'll try to stick to it. Changed.

File: drivers/watchdog/wdt_gecko.c [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+		}
+
+		/* Found the (rounded up?) value of cycles for
+		 *  supplied timeout
+		 */
+		break;
+	}
+
+	return idx;
+}
```
Comment (mnkp) [inline drivers/watchdog/wdt_gecko.c]: I propose to write this part as:
```
/* Defines maximum WDOG_CTRL.PERSEL value which is used by the watchdog module
 * to select its timeout period.
 */
#define WDT_GECKO_MAX_PERIOD_SELECT_VALUE 15

static u32_t wdt_gecko_get_timeout_from_persel(int perSel)
{
	return (8 << perSel) + 1;
}

/* Find the rounded up value of cycles for supplied timeout. When using ULFRCO
 * (default), 1 cycle is 1 ms +/- 12%.
 */
static int wdt_gecko_get_persel_from_timeout(u32_t timeout)
{
	int idx;

	for (idx = 0; idx < WDT_GECKO_MAX_PERIOD_SELECT_VALUE; idx++) {
		if (wdt_gecko_get_timeout_from_persel(idx) >= timeout) {
			break;
		}
	}

	return idx;
}
```
and remove `timeout_in_cycles_tbl` table. This implementation should reduce driver code size.

File: drivers/watchdog/wdt_gecko.c [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+
+	if ((cfg->window.max < timeout_in_cycles_tbl[0]) ||
+		(cfg->window.max > timeout_in_cycles_tbl[
+			ARRAY_SIZE(timeout_in_cycles_tbl) - 1])) {
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
```
Comment (mnkp) [inline drivers/watchdog/wdt_gecko.c]: The actual timeout period set by the watchdog will be widely inaccurate compared to the required value. E.g. given 131074 timeout value provided via the API call the watchdog will expire after 262145. As such there is no reason to reject value 262146. I propose to remove this check entirely. If we are concerned about the accuracy of the timeout setting we need to extend the watchdog API. E.g. provide an extra function that will allow application to read out the timeout value set by the watchdog.

**Review by mnkp — CHANGES_REQUESTED** [review]
@Oanerer Thanks, that looks good! I do have however a few more comments I overlooked during the last review.

File: drivers/watchdog/wdt_gecko.c [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+	data->wdog_config.em2Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+	data->wdog_config.em3Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+
+	data->wdog_config.debugRun =
+		(options & WDT_OPT_PAUSE_HALTED_BY_DBG) == 0U;
+
+	if (data->callback != NULL) {
+		/* Interrupt mode for window */
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: I'll add it for completeness sake, but with the Zephyr initialization sequence of drivers I can't really think of a scenario where the status register of the watchdog interrupt would be 'dirty' upon initialization of this driver.

File: dts/arm/silabs/efm32pg12b.dtsi [hunk 13]
```diff
@@ -10,6 +10,11 @@
...
+		watchdog0 = &wdog0;
+		watchdog1 = &wdog1;
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply dts/arm/silabs/efm32pg12b.dtsi]: I was explicitly told by Galak to add the aliases to the SoC DTS file. The [original PR](https://github.com/zephyrproject-rtos/zephyr/pull/20321/commits/d60d3ab0f262c0e8273d61350904c14286f06d8e) that added the aliases in the first place also added them there for the Atmel and NXP drivers. As far as I can see they are not alias based right?

I'll happily change this to board DTS if necessary, but I would like confirmation first that they indeed belong there and that this is in line with the other watchdog drivers.

File: drivers/watchdog/wdt_gecko.c [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+
+#ifdef CONFIG_WDT_DISABLE_AT_BOOT
+	/* Ignore any errors */
+	wdt_gecko_disable(dev);
+#endif
+
+	/* Enable ULFRCO (1KHz) oscillator */
+	CMU_OscillatorEnable(cmuOsc_ULFRCO, true, false);
+
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Nice catch. I can't really explain either why it works on my Pearl Gecko? I'll add this.

File: drivers/watchdog/wdt_gecko.c [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+static int wdt_gecko_install_timeout(struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_Init_TypeDef init_defaults = WDOG_INIT_DEFAULT;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: The same variable name is used across the Atmel and NXP drivers, which I used as a 'template'. So I chose to stick with uniformity. If you think ``timeout_installed`` is a clearer description, I'll change it.

File: drivers/watchdog/wdt_gecko.c [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+			ARRAY_SIZE(timeout_in_cycles_tbl) - 1])) {
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
+
+	data->wdog_config = init_defaults;
+
+	data->wdog_config.perSel = (WDOG_PeriodSel_TypeDef)
+		wdt_gecko_convert_timeout(cfg->window.max);
+
```
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Sounds reasonable

File: drivers/watchdog/wdt_gecko.c (hunk 10 above)
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: I'll change it, looks cleaner.

File: dts/arm/silabs/efm32pg12b.dtsi (hunk 13 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply dts/arm/silabs/efm32pg12b.dtsi]: As I mentioned, I believe we shouldn't have an alias section in the SoC DTS files on SiLabs but let's wait for @galak in this case. He's responsible for DTS, has a better overview.

File: drivers/watchdog/wdt_gecko.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/watchdog/wdt_gecko.c]: I would change it, but if you don't want I'm not going to interfere.

File: drivers/watchdog/wdt_gecko.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/watchdog/wdt_gecko.c]: Actually, we only need to clear the status before enabling the interrupt. If we disable it it's not necessary, but doesn't hurt either.

File: drivers/watchdog/wdt_gecko.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/watchdog/wdt_gecko.c]: we could also print the unit `ms`.

**Comment by Oanerer** [comment]
I'm still confused on how the watchdog test at ``tests/drivers/watchdog/wdt_basic_api`` is supposed to work. 

The first test ``test_wdt_no_callback()`` installs a timeout value, does not set a callback function and sets flag ``WDT_FLAG_RESET_SOC``. This simply lets the MCU reset when the timeout is reached. Working as intended.

However, the second test ``test_wdt_callback_1()`` also installs a timeout value, sets a callback value but also still sets flag ``WDT_FLAG_RESET_SOC``. This means that the reset takes precedence over the callback function, so the MCU immediately resets without calling the callback function. The result is that the second test fails on my hardware:
```
Testcase: test_wdt_callback_1

    Assertion failed at ZEPHYR_BASE/tests/drivers/watchdog/wdt_basic_api/src/test_wdt.c:312: test_wdt: test_wdt_callback_1() == TC_PASS is false

FAIL - test_wdt
```
In a previous iteration of the driver I used the early warning interrupt to call the installed callback function at 75% of the timeout value and then let the reset do it's thing at the timeout value, but apparently this was not how it's supposed to work:

@mnkp wrote:
> 
> 
> If we want to have different things happening on the same watchdog channel at different times, e.g. interrupt at 75% and reset at 100% we need to use the concept of multi-stage watchdog. This should be supported only if `CONFIG_WDT_MULTISTAGE` is enabled. Otherwise the callback should be triggered when timeout expires. This is controlled by bit 0 (TOUT Interrupt Enable) in `WDOG_IEN` register.
> 
> Of course, having a callback triggered upon timeout expire event makes sense only if `WDT_FLAG_RESET_NONE` is passed. We probably don't need to take this into account when writing the driver since reset will take precedence.

Could someone explain to me how this test is supposed to work so I can adjust my driver to function accordingly? Or should ``test_wdt_callback_1()`` not use ``WDT_FLAG_RESET_SOC`` but ``WDT_FLAG_RESET_NONE`` instead?

**Comment by mnkp** [comment]
@Oanerer Thanks for pointing out the inconsistencies. From my perspective the testcase doesn't comply with the API specification. The API was not designed to behave the way testcase is expecting it to. This needs to be clarified however. It would be best if you opened a GH issue describing shortly your findings.

**Comment by Oanerer** [comment]
@mnkp: thank you, I will do that!

File: drivers/watchdog/wdt_gecko.c [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+	} else {
+		/* Normal mode */
+		data->wdog_config.winSel = wdogIllegalWindowDisable;
+	}
+
+	/* Set mode of watchdog and callback */
+	switch (cfg->flags) {
+	case WDT_FLAG_RESET_SOC:
+	case WDT_FLAG_RESET_CPU_CORE:
+		data->wdog_config.resetDisable = false;
```
Comment (pabigot) [inline drivers/watchdog/wdt_gecko.c]: #20693 suggests this platform does not provide a delay in which a callback can be invoked before the reset occurs.  If so I believe the documented API would require that this section return `-ENOTSUP` if a callback was passed.

Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Thank you for insight! I changed the driver to reflect the -ENOTSUP status. Please let me know what you think.

File: dts/arm/silabs/efm32pg12b.dtsi (hunk 13 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply dts/arm/silabs/efm32pg12b.dtsi]: @galak could you have a look at this. The issue discussion above.

[the remark this answers is not in the document]
Comment (Oanerer) [reply dts/arm/silabs/efm32pg12b.dtsi]: @galak: could you take a look at this please when you have the time?

[the remark this answers is not in the document]
Comment (galak) [reply dts/arm/silabs/efm32pg12b.dtsi]: In discussion with @mnkp I'm ok with this in the board dts, we just need a comment in the board dts that the aliases are specifically for the watchdog samples.

[the remark this answers is not in the document]
Comment (Oanerer) [reply dts/arm/silabs/efm32pg12b.dtsi]: Thanks guys for the help! I've moved the aliases to the board dts file along with the usual `/* These aliases are provided for compatibility with samples */` comment to indicate what they are for.

File: drivers/watchdog/wdt_gecko.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Modified to only clear possible lingering interrupts before enabling interrupts

File: drivers/watchdog/wdt_gecko.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (Oanerer) [reply drivers/watchdog/wdt_gecko.c]: Renamed to `timeout_installed`

**Review by mnkp — APPROVED** [review]

File: dts/bindings/watchdog/silabs,gecko-wdog.yaml [hunk 18]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2019 Interay Solutions B.V.
+# Copyright (c) Oane Kingma
+# SPDX-License-Identifier: Apache-2.0
+
+title: Silicon Labs Gecko Family Watchdog driver
```
Comment (MaureenHelm) [inline dts/bindings/watchdog/silabs,gecko-wdog.yaml]: Please remove the title and update the description per #20793

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
One nit, otherwise looks good.

File: dts/bindings/watchdog/silabs,gecko-wdog.yaml (hunk 18 above)
Comment (Oanerer) [reply dts/bindings/watchdog/silabs,gecko-wdog.yaml]: Fixed

**Comment by Oanerer** [comment]
Thanks Maureen! Minor nit fixed.

**Review by MaureenHelm — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**